src/test/ui/consts/const_discriminant.rs

```diff
@@ -0,0 +1,22 @@
+// run-pass
+#![feature(const_discriminant)]
+
+use std::mem::{discriminant, Discriminant};
+
+enum Test {
+    A(u8),
+    B,
+    C { a: u8, b: u8 },
+}
+
+const TEST_A: Discriminant<Test> = discriminant(&Test::A(5));
+const TEST_A_OTHER: Discriminant<Test> = discriminant(&Test::A(17));
+const TEST_B: Discriminant<Test> = discriminant(&Test::B);
+
+fn main() {
+    assert_eq!(TEST_A, TEST_A_OTHER);
+    assert_eq!(TEST_A, discriminant(&Test::A(17)));
```

Comment:
This might get promoted in the future (I hope not but who knows), in which case we'd just compare const-eval against itself here.
What some other tests do is to run the result through an identity function:
```
fn id<T>(x: T) { x}

assert_eq!(TEST_A, discriminant(id(&Test::A(17))));
```

Reply:
ok.

I still have some questions about this though:

1. I did not use `convert::identity` as it is also const, meaning that it would probably also get promoted, is this assumption correct?.
2. is there a difference between getting promoted and *just* getting executed at compile time?
  2.1 In case there is none: *This might get promoted in the future (I hope not but who knows)*: Wouldn't it be a good thing if `discriminant(/* const expr */)` gets promoted. I sometimes use
`mem::discriminant(&s.action) == mem::discriminant(&Action::LoadInput(0))` and expect the rhs to compile to a constant.

Reply:
Promotion and constant-propagating are not the same thing. The latter is an optimization that does not change observable behavior. [Promotion](https://github.com/rust-lang/rfcs/blob/master/text/1414-rvalue_static_promotion.md) even affects type-checking.

To be fair, const-propagation also uses Miri to some extend, but IIRC passing things through an identity function was considered good enough... @oli-obk please correct me, you convinced me we had to do this.^^

> I did not use convert::identity as it is also const, meaning that it would probably also get promoted, is this assumption correct?

Well, it is the cautious thing to do. :)

Reply:
Ok, using a custom identity function prevents promotion as promoting `non_const_fn(const_expr)` is not allowed.

Still not sure how safe this is in the long term, as I expect that calling an intrinsic during
const propagation probably also uses miri. :man_shrugging: